kernel/src/drivers/keyboard/mod.rs

```diff
@@ -0,0 +1,200 @@
+//! The event contains the keycode pressed, which can be compared to `keymap::codes`, an optional `char`, the type of press, and various modifier flags.
+
+pub mod keymap;
+
+use core::convert::From;
+
+use drivers::ps2::{self, Device, DeviceState};
+use drivers::ps2::io::Ps2Error;
+use drivers::ps2::io::commands::{DeviceCommand, DeviceDataCommand};
+
+/// Interface to a generic keyboard.
+pub trait Keyboard {
+    type Error;
+
+    /// Enables this keyboard, setting it up before use.
+    ///
+    /// # Note
+    ///
+    /// The keyboard should not be accessed while not enabled!
+    fn enable(&mut self) -> Result<(), Self::Error>;
+
+    /// Disables this keyboard, making use unavailable.
+    ///
+    /// # Note
+    ///
+    /// Until `enable` is called again, this keyboard should not be used
+    fn disable(&mut self) -> Result<(), Self::Error>;
+
+    /// Polls the device for state events.
+    /// TODO: This should eventually use interrupts and hold a queue
+    fn read_event(&mut self) -> Result<KeyEvent, Self::Error>;
+
+    /// Returns true if the given keycode is currently being pressed
+    fn pressed(&self, keycode: u8) -> bool;
+}
+
+/// Contains data relating to a key press event.
+#[derive(Copy, Clone, Debug)]
+pub struct KeyEvent {
+    pub keycode: u8,
+    pub char: Option<char>,
+    pub event_type: KeyEventType,
+    pub ctrl: bool,
+    pub alt: bool,
+    pub shift: bool,
+}
+
+/// The type of key event occurring
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+pub enum KeyEventType {
+    Make,
+    Break,
+    Repeat,
+}
+
+/// An error for a PS/2 keyboard
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+pub enum Ps2KeyboardError {
+    ReadError(Ps2Error),
+    KeyboardDisabled,
+    ScancodeNotSet(u8),
+    ScanningNotEnabled,
+    DataUnavailable,
+}
+
+/// Handles interface to a PS/2 keyboard, if available
+pub struct Ps2Keyboard<'a> {
+    device: &'a mut Device,
+    key_states: [bool; 0xFF],
+}
+
+impl<'a> Ps2Keyboard<'a> {
+    pub fn new(device: &'a mut Device) -> Self {
+        Ps2Keyboard {
+            device,
+            key_states: [false; 0xFF],
+        }
+    }
+
+    /// Reads a single scancode from this
+    fn read_scancode(&self) -> Result<Ps2Scancode, Ps2KeyboardError> {
+        self.check_enabled()?;
+
+        if ps2::io::can_read()? && ps2::io::can_read_keyboard()? {
+            // Check scancode present
+            let scancode = ps2::io::read(&ps2::io::DATA_PORT)?;
+
+            // Check not extended scancode (ignored)
+            if scancode != 0 && scancode != 0xE0 && scancode != 0xE1 {
+                // If not key release (break) code
+                if scancode < 0x80 {
+                    return Ok(Ps2Scancode::new(scancode, true));
+                } else {
+                    return Ok(Ps2Scancode::new(scancode - 0x80, false));
+                }
+            }
+        }
+
+        Err(Ps2KeyboardError::DataUnavailable)
```

Comment:
Is this necessarily an error? It could be encoded into an Option as `None`

Reply:
Yes, `Ps2Error`s from `read` are translated to `Ps2KeyboardError`s.

Reply:
Yes, but it could return Ok(None), no?

Reply:
No. The error should be propagated through for callers. If an error occurs related to PS/2 when calling this, you'd want to know about it.

Reply:
This isn't related to PS/2 other than it just doesn't have any data for you, though?

Reply:
No, the PS/2 controller might throw a different error (although none yet) which is worth nothing.

Reply:
The fact that the code has got here already means that it *hasn't* thrown an error. It just has nothing for you, which should be a `None`, right?

Reply:
Ah, I misunderstood your question. Anyway, the nesting of `Result<Option<_>, _>` seems a bit over the top.

Reply:
Eh... It's not the worst I've seen. IMO, we should do it.